src/hotspot/share/gc/g1/g1FullCollector.cpp

```diff
@@ -207,9 +209,15 @@ void G1FullCollector::collect() {
 
   phase2_prepare_compaction();
 
-  phase3_adjust_pointers();
+  if (has_compaction_targets()) {
+    phase3_adjust_pointers();
 
-  phase4_do_compaction();
+    phase4_do_compaction();
+  } else {
+    log_info(gc, phases) ("No Regions selected for compaction. Skipping Phase 3: Adjust pointers and Phase 4: Compact heap");
```

Comment:
A comment on when this will be the case would be nice. As I understand it, it's `all regions are almost full & max-compaction == false`.